File: apps/web/src/lib/api/dispersion.ts
/**
 * Atmospheric Dispersion & Downwind Hazard Intelligence API client for SIH26162 (Phase 3 & 4).
 */

import { apiFetch } from "./client";
import type {
  AtmosphereData,
  Coordinate,
  DataQuality,
  WindVector,
} from "@/types/weather";
import type {
  AtmosphericDispersionResult,
  DispersionCalculationResponse,
  DispersionSamplePoint,
  DispersionSummary,
  PasquillStabilityClass,
} from "@/types/dispersion";

export type {
  AtmosphericDispersionResult,
  DispersionCalculationResponse,
  DispersionSamplePoint,
  DispersionSummary,
  PasquillStabilityClass,
};

export interface FetchDispersionOptions {
  frpMw?: number;
  releaseHeightM?: number;
  customWindSpeedMs?: number;
  customWindDirectionDeg?: number;
  isDaytime?: boolean;
  maxDistanceKm?: number;
  signal?: AbortSignal;
}

export interface DispersionCalculationRequestBody {
  latitude: number;
  longitude: number;
  frp_mw?: number | null;
  release_height_m?: number | null;
  custom_wind_speed_ms?: number | null;
  custom_wind_direction_deg?: number | null;
  is_daytime?: boolean | null;
  max_distance_km?: number | null;
}

/**
 * Fetch atmospheric dispersion hazard profile using GET query parameters.
 */
export async function fetchDispersion(
  latitude: number,
  longitude: number,
  options: FetchDispersionOptions = {}
): Promise<DispersionCalculationResponse> {
  const {
    frpMw,
    releaseHeightM,
    customWindSpeedMs,
    customWindDirectionDeg,
    isDaytime,
    maxDistanceKm,
    signal,
  } = options;

  return apiFetch<DispersionCalculationResponse>("/dispersion", {
    params: {
      latitude,
      longitude,
      frp_mw: frpMw,
      release_height_m: releaseHeightM,
      custom_wind_speed_ms: customWindSpeedMs,
      custom_wind_direction_deg: customWindDirectionDeg,
      is_daytime: isDaytime,
      max_distance_km: maxDistanceKm,
    },
    signal,
  });
}

/**
 * Calculate atmospheric dispersion via POST body with explicit parameters.
 */
export async function calculateDispersion(
  body: DispersionCalculationRequestBody,
  signal?: AbortSignal
): Promise<DispersionCalculationResponse> {
  return apiFetch<DispersionCalculationResponse>("/dispersion", {
    method: "POST",
    body: JSON.stringify(body),
    headers: {
      "Content-Type": "application/json",
    },
    signal,
  });
}

/**
 * Fetch atmospheric dispersion hazard profile coupled to a specific thermal event.
 */
export async function fetchEventDispersion(
  eventId: string,
  latitude: number,
  longitude: number,
  options: FetchDispersionOptions = {}
): Promise<DispersionCalculationResponse> {
  const { frpMw, releaseHeightM, maxDistanceKm, signal } = options;

  return apiFetch<DispersionCalculationResponse>(
    `/dispersion/events/${encodeURIComponent(eventId)}`,
    {
      params: {
        latitude,
        longitude,
        frp_mw: frpMw,
        release_height_m: releaseHeightM,
        max_distance_km: maxDistanceKm,
      },
      signal,
    }
  );
}

File: services/api/routes/dispersion.py
"""FastAPI route handlers for Atmospheric Dispersion & Downwind Hazard Intelligence (Phase 3)."""

from typing import Annotated

from fastapi import APIRouter, Body, Depends, Query, status

from packages.config.settings import Settings
from packages.data.weather import get_dispersion_service
from packages.errors import ValidationError
from services.api.dependencies import get_app_settings
from services.api.schemas.dispersion import (
    DispersionCalculationRequest,
    DispersionCalculationResponse,
)

router = APIRouter(prefix="/dispersion", tags=["dispersion"])


@router.post(
    "",
    response_model=DispersionCalculationResponse,
    status_code=status.HTTP_200_OK,
    summary="Compute atmospheric dispersion and hazard corridor for coordinates",
    description=(
        "Performs Gaussian plume atmospheric dispersion modeling using validated meteorological "
        "conditions (or custom parameter overrides). Evaluates downwind centerline trajectory, "
        "lateral cross-sections, Pasquill-Gifford stability, and relative ground-level concentration decay."
    ),
)
def calculate_dispersion_post(
    settings: Annotated[Settings, Depends(get_app_settings)],
    request: Annotated[
        DispersionCalculationRequest,
        Body(description="Atmospheric dispersion calculation parameters"),
    ],
) -> DispersionCalculationResponse:
    """Calculate atmospheric dispersion from JSON request body."""
    dispersion_service = get_dispersion_service(settings)
    result = dispersion_service.calculate_dispersion(
        latitude=request.latitude,
        longitude=request.longitude,
        frp_mw=request.frp_mw,
        release_height_m=request.release_height_m,
        custom_wind_speed_ms=request.custom_wind_speed_ms,
        custom_wind_direction_deg=request.custom_wind_direction_deg,
        is_daytime=request.is_daytime,
        max_distance_km=request.max_distance_km,
    )

    return DispersionCalculationResponse(
        source_location=result.source_location,
        event_id=result.event_id,
        evaluated_at=result.evaluated_at,
        wind=result.wind,
        dispersion=result.dispersion,
        trajectory=result.trajectory,
        data_quality=result.data_quality,
        model_confidence=result.model_confidence,
    )


@router.get(
    "",
    response_model=DispersionCalculationResponse,
    status_code=status.HTTP_200_OK,
    summary="Compute atmospheric dispersion and hazard corridor (GET query parameters)",
    description=(
        "Query-based Gaussian plume atmospheric dispersion calculation for rapid UI inspection "
        "and downstream subsystem integration."
    ),
)
def calculate_dispersion_get(
    settings: Annotated[Settings, Depends(get_app_settings)],
    latitude: Annotated[
        float | None,
        Query(ge=-90.0, le=90.0, description="Latitude in decimal degrees [-90, 90]"),
    ] = None,
    longitude: Annotated[
        float | None,
        Query(ge=-180.0, le=180.0, description="Longitude in decimal degrees [-180, 180]"),
    ] = None,
    lat: Annotated[
        float | None,
        Query(ge=-90.0, le=90.0, description="Alias for latitude"),
    ] = None,
    lon: Annotated[
        float | None,
        Query(ge=-180.0, le=180.0, description="Alias for longitude"),
    ] = None,
    frp_mw: Annotated[
        float | None,
        Query(ge=0.0, description="Fire Radiative Power in MW"),
    ] = None,
    release_height_m: Annotated[
        float | None,
        Query(ge=0.0, description="Stack / emission release height in meters"),
    ] = None,
    custom_wind_speed_ms: Annotated[
        float | None,
        Query(ge=0.0, description="Custom wind speed override in m/s"),
    ] = None,
    custom_wind_direction_deg: Annotated[
        float | None,
        Query(ge=0.0, le=360.0, description="Custom wind direction override in degrees"),
    ] = None,
    is_daytime: Annotated[
        bool | None,
        Query(description="Solar daytime flag override"),
    ] = None,
    max_distance_km: Annotated[
        float | None,
        Query(ge=0.5, le=50.0, description="Downwind calculation horizon in km"),
    ] = None,
) -> DispersionCalculationResponse:
    """Calculate atmospheric dispersion via query parameters."""
    eff_lat = latitude if latitude is not None else lat
    eff_lon = longitude if longitude is not None else lon

    if eff_lat is None or eff_lon is None:
        raise ValidationError(
            "Both latitude (or 'lat') and longitude (or 'lon') query parameters are required.",
            details={
                "provided_latitude": eff_lat,
                "provided_longitude": eff_lon,
            },
        )

    dispersion_service = get_dispersion_service(settings)
    result = dispersion_service.calculate_dispersion(
        latitude=eff_lat,
        longitude=eff_lon,
        frp_mw=frp_mw,
        release_height_m=release_height_m,
        custom_wind_speed_ms=custom_wind_speed_ms,
        custom_wind_direction_deg=custom_wind_direction_deg,
        is_daytime=is_daytime,
        max_distance_km=max_distance_km,
    )

    return DispersionCalculationResponse(
        source_location=result.source_location,
        event_id=result.event_id,
        evaluated_at=result.evaluated_at,
        wind=result.wind,
        dispersion=result.dispersion,
        trajectory=result.trajectory,
        data_quality=result.data_quality,
        model_confidence=result.model_confidence,
    )


@router.get(
    "/events/{event_id}",
    response_model=DispersionCalculationResponse,
    status_code=status.HTTP_200_OK,
    summary="Compute atmospheric dispersion coupled to a thermal event",
    description="Evaluates atmospheric dispersion and hazard corridor for a specific thermal event coordinate.",
)
def get_event_dispersion(
    event_id: str,
    settings: Annotated[Settings, Depends(get_app_settings)],
    latitude: Annotated[
        float | None,
        Query(ge=-90.0, le=90.0, description="Event latitude coordinate"),
    ] = None,
    longitude: Annotated[
        float | None,
        Query(ge=-180.0, le=180.0, description="Event longitude coordinate"),
    ] = None,
    frp_mw: Annotated[
        float | None,
        Query(ge=0.0, description="Fire Radiative Power in MW"),
    ] = None,
    release_height_m: Annotated[
        float | None,
        Query(ge=0.0, description="Release height in meters"),
    ] = None,
    max_distance_km: Annotated[
        float | None,
        Query(ge=0.5, le=50.0, description="Downwind horizon in km"),
    ] = None,
) -> DispersionCalculationResponse:
    """Evaluate atmospheric dispersion for a given thermal anomaly event."""
    from services.api.services.events import EventQueryService

    eff_lat = latitude
    eff_lon = longitude
    eff_frp = frp_mw

    if eff_lat is None or eff_lon is None:
        event_detail = EventQueryService.get_event(event_id)
        lon, lat = event_detail.geometry["coordinates"]
        eff_lat = eff_lat if eff_lat is not None else lat
        eff_lon = eff_lon if eff_lon is not None else lon

    dispersion_service = get_dispersion_service(settings)
    result = dispersion_service.evaluate_event_dispersion(
        event_id=event_id,
        latitude=eff_lat,
        longitude=eff_lon,
        frp_mw=eff_frp,
        release_height_m=release_height_m,
        max_distance_km=max_distance_km,
    )

    return DispersionCalculationResponse(
        source_location=result.source_location,
        event_id=result.event_id,
        evaluated_at=result.evaluated_at,
        wind=result.wind,
        dispersion=result.dispersion,
        trajectory=result.trajectory,
        data_quality=result.data_quality,
        model_confidence=result.model_confidence,
    )
